sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -1745,4 +1765,246 @@ def get_transceiver_loopback(self):
+                Integer, a bitmask of the lanes on the host side
+                e.g. 0x5 for lane 0 and lane 2.
+
+        Returns:
+            (has_update, new_appl)
+        """
+        if host_speed == 0 or host_lanemask == 0 or self.is_flat_memory():
+            return (False, 1)
+
+        app_new = self.get_cmis_application_matched(host_speed, host_lanemask)
+        if app_new != 1 or host_lanemask != (1 << self.NUM_CHANNELS) - 1:
+            logger.info("Non-default application is not supported")
+            return (False, 1)
+
+        app_old = 0
+        for lane in range(self.NUM_CHANNELS):
+            if (1 << lane) & host_lanemask == 0:
+                continue
+            if app_old == 0:
+                app_old = self.get_cmis_application_selected(lane)
+            elif app_old != self.get_cmis_application_selected(lane):
+                logger.info("Not all the lanes are in the same application mode")
+                logger.info("Forcing application update...")
+                return (True, app_new)
+
+        if app_old == app_new:
+            skip = True
+            dp_state = self.get_datapath_state()
+            conf_state = self.get_config_datapath_hostlane_status()
+            for lane in range(self.NUM_CHANNELS):
+                if (1 << lane) & host_lanemask == 0:
+                    continue
+                name = "DP{}State".format(lane + 1)
+                if dp_state[name] != CmisCodes.DATAPATH_STATE[4]:
+                    skip = False
+                    break
+                name = "ConfigStatusLane{}".format(lane + 1)
+                if conf_state[name] != CmisCodes.CONFIG_STATUS[1]:
+                    skip = False
+                    break
+            if skip:
+                return (False, app_old)
+
+        return (True, app_new)
+
+    def set_cmis_application_apsel(self, host_lanemask, appl_code):
+        """
+        Update the selected application code to the specified host lanes
+
+        Args:
+            host_lanemask:
+                Integer, a bitmask of the lanes on the host side
+                e.g. 0x5 for lane 0 and lane 2.
+            appl_code:
+                Integer, the desired application code
+
+        Returns:
+            Boolean, true if success otherwise false
+        """
+        # Update the application selection
+        for lane in range(self.NUM_CHANNELS):
+            if (1 << lane) & host_lanemask == 0:
+                continue
+            addr = "{}_{}_{}".format(consts.STAGED_CTRL_APSEL_FIELD, 0, lane + 1)
+            data = appl_code << 4
+            self.xcvr_eeprom.write(addr, data)
+
+        # Apply DataPathInit
+        return self.xcvr_eeprom.write("%s_%d" % (consts.STAGED_CTRL_APPLY_DPINIT_FIELD, 0), host_lanemask)
+
+    def get_error_description(self):
+        dp_state = self.get_datapath_state()
+        conf_state = self.get_config_datapath_hostlane_status()
+        for lane in range(self.NUM_CHANNELS):
```

Comment:
can you define a function (get_cmis_num_channels()) to get the count?. Optics like DR4 has only 4 channels. we can't assume 8 as default number of channels.

Reply:
Done, but please note these are host lanes connected to the MAC/ASIC, hence it's 8 lanes in DR4, as to QSFP+ CMIS, it's most likely 4 lanes.

e.g. Here is an example of INNOLIGHT 400G DR4, the ConfigError reside at 0x94a, while DPState are at 0x900
```
root@sonic:~# hexdump -C -n 256 -s 0x880 /sys/bus/i2c/devices/40-0050/eeprom
00000880  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000890  00 21 21 25 25 29 29 2d  2d ff 00 00 33 33 33 33  |.!!%%))--...3333|
000008a0  ff ff 22 22 22 22 00 00  00 00 22 22 22 22 00 00  |..""""....""""..|
000008b0  00 00 00 00 10 10 10 10  10 10 10 10 ff 00 00 33  |...............3|
000008c0  33 33 33 ff ff 22 22 22  22 00 00 00 00 22 22 22  |333..""""...."""|
000008d0  22 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |"...............|
000008e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000900  44 44 44 44 00 00 00 00  00 00 00 00 00 00 00 00  |DDDD............|
00000910  00 00 00 00 00 00 00 00  00 00 3e db 3e 51 3c ca  |..........>.>Q<.|
00000920  3d 09 00 00 00 00 00 00  00 00 9c 40 9c 40 9c 40  |=..........@.@.@|
00000930  9c 40 00 00 00 00 00 00  00 00 3a 7c 56 23 31 23  |.@........:|V#1#|
00000940  32 7e 00 00 00 00 00 00  00 00 11 11 11 11 21 21  |2~............!!|
00000950  25 25 29 29 2d 2d ff 00  00 33 33 33 33 ff ff 22  |%%))--...3333.."|
00000960  22 22 22 00 00 00 00 22  22 22 22 00 00 00 00 00  |"""...."""".....|
00000970  11 12 13 14 00 00 00 00  11 12 13 14 00 00 00 00  |................|
```